sql/parser/eval.go

```diff
@@ -1214,11 +1434,61 @@ func evalCastExpr(expr *CastExpr) (Datum, error) {
 		}
 		return s, nil
 
-		// TODO(pmattis): unimplemented.
-		// case *DecimalType:
-		// case *DateType:
-		// case *TimestampType:
+	// TODO(pmattis): unimplemented.
+	// case *DecimalType:
+	case *DateType:
+		switch d.(type) {
+		case DString:
+			return ParseDate(d.(DString))
+		}
+	case *TimestampType:
+		switch d.(type) {
+		case DString:
+			return ParseTimestamp(d.(DString))
+		}
+	case *IntervalType:
+		switch d.(type) {
+		case DString:
+			d, err := time.ParseDuration(string(d.(DString)))
+			return DInterval{Duration: d}, err
+		}
 	}
 
 	return DNull, fmt.Errorf("invalid cast: %s -> %s", d.Type(), expr.Type)
 }
+
+const (
+	dateFormat                    = "2006-01-02"
+	timestampFormat               = "2006-01-02 15:04:05.999999999"
+	timestampWithOffsetZoneFormat = "2006-01-02 15:04:05.999999999-07:00"
+	timestampWithNamedZoneFormat  = "2006-01-02 15:04:05.999999999 MST"
+)
+
+// ParseDate parses a date.
+func ParseDate(s DString) (DDate, error) {
+	str := string(s)
+	t, err := time.Parse(dateFormat, str)
+	if err == nil {
+		return DDate{Time: t}, nil
+	}
+	return DDate{}, err
+}
+
+// ParseTimestamp parses the timestamp.
+func ParseTimestamp(s DString) (DTimestamp, error) {
+	str := string(s)
+	t, err := time.Parse(timestampFormat, str)
+	if err == nil {
+		return DTimestamp{Time: t}, nil
+	}
+	t, err = time.Parse(timestampWithOffsetZoneFormat, str)
+	if err == nil {
+		return DTimestamp{Time: t}, nil
+	}
+	t, err = time.Parse(timestampWithNamedZoneFormat, str)
+	if err == nil {
+		// This is more involved. See https://github.com/golang/go/issues/12388
```

Comment:
Seems like the bug you're pointing to only affects the Go playground. Am I missing something?